metaflow/subprocess_manager.py

```diff
@@ -0,0 +1,278 @@
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.cleanup()
+
+    def handle_sigint(self, signum, frame):
+        print("SIGINT received.")
+        asyncio.create_task(self.kill())
+
+    async def wait(self, timeout=None, stream=None):
+        if timeout is None:
+            if stream is None:
+                await self.process.wait()
+            else:
+                await self.emit_logs(stream)
+        else:
+            tasks = [asyncio.create_task(asyncio.sleep(timeout))]
+            if stream is None:
+                tasks.append(asyncio.create_task(self.process.wait()))
+            else:
+                tasks.append(asyncio.create_task(self.emit_logs(stream)))
+
+            await asyncio.wait(tasks, return_when="FIRST_COMPLETED")
+
+    async def run(self):
+        self.temp_dir = tempfile.mkdtemp()
+        stdout_logfile = os.path.join(self.temp_dir, "stdout.log")
+        stderr_logfile = os.path.join(self.temp_dir, "stderr.log")
+
+        try:
+            # returns when process has been started,
+            # not when it is finished...
+            self.process = await asyncio.create_subprocess_exec(
+                *self.command,
+                cwd=self.cwd,
+                env=self.env,
+                stdout=open(stdout_logfile, "w"),
+                stderr=open(stderr_logfile, "w"),
+            )
+
+            self.log_files["stdout"] = stdout_logfile
+            self.log_files["stderr"] = stderr_logfile
+
+            self.run_called = True
+            return self.process
+        except Exception as e:
+            print(f"Error starting subprocess: {e}")
+            await self.cleanup()
+
+    async def stream_logs(
+        self, stream, position=None, timeout_per_line=None, log_write_delay=0.01
+    ):
+        if self.run_called is False:
+            raise ValueError("No command run yet to get the logs for...")
+
+        if stream not in self.log_files:
+            raise ValueError(
+                f"No log file found for {stream}, valid values are: {list(self.log_files.keys())}"
+            )
+
+        log_file = self.log_files[stream]
+
+        with open(log_file, mode="r") as f:
+            if position is not None:
+                f.seek(position)
+
+            while True:
+                # wait for a small time for complete lines to be written to the file
+                # else, there's a possibility that a line may not be completely written when
+                # attempting to read it. This is not a problem, but improves readability.
+                await asyncio.sleep(log_write_delay)
+
+                try:
+                    if timeout_per_line is None:
+                        line = f.readline()
+                    else:
+                        line = await asyncio.wait_for(f.readline(), timeout_per_line)
+                except asyncio.TimeoutError as e:
+                    raise LogReadTimeoutError(
```

Comment:
we could also return the position and None?

Reply:
So we would return:
- the actual line
- an empty line if EOF or just an actual empty line
- None if a timeout.

I don't know if we want to distinguish the cases in 2.

Reply:
can discuss for more clarity...